kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/engine/DefaultJsonHandler.java

```diff
@@ -117,18 +117,22 @@ public boolean hasNext() {
         // read.
         try {
           if (currentFileReader == null || (nextLine = currentFileReader.readLine()) == null) {
-
-            tryOpenNextFile();
-            if (currentFileReader != null) {
-              nextLine = currentFileReader.readLine();
+            // `nextLine` will initially be null because `currentFileReader` is guaranteed
+            // to be null
+            while (nextLine == null) {
+              tryOpenNextFile();
+              if (currentFileReader != null) {
+                nextLine = currentFileReader.readLine();
+              } else {
+                return false;
+              }
```

Comment:
```suggestion
            // `nextLine` will initially be null because `currentFileReader` is guaranteed
            // to be null
            while (nextLine == null) {
              if (tryOpenNextFile()) {
                 return hasNext();
              }
            }
```

Reply:
basically modify the `tryOpenNextFile` to return true if a new file is opened, false otherwise. Depending upon that call the hasNext(). Simplifies the logic.